<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN""http://www.w3.org/TR/html4/loose.dtd">
<HTML
><HEAD
><TITLE
>Exceptions</TITLE
><META
CHARSET="UTF-8"><META
NAME="GENERATOR"
CONTENT="Modular DocBook HTML Stylesheet Version 1.79"><LINK
REL="HOME"
TITLE="ATSプログラミングチュートリアル"
HREF="index.html"><LINK
REL="UP"
TITLE="基本的なチュートリアル"
HREF="p37.html"><LINK
REL="PREVIOUS"
TITLE="Fixity Declarations"
HREF="c93.html"><LINK
REL="NEXT"
TITLE="References"
HREF="c194.html"></HEAD
><BODY
CLASS="chapter"
BGCOLOR="#FFFFFF"
TEXT="#000000"
LINK="#0000FF"
VLINK="#840084"
ALINK="#0000FF"
><DIV
CLASS="NAVHEADER"
><TABLE
SUMMARY="Header navigation table"
WIDTH="100%"
BORDER="0"
CELLPADDING="0"
CELLSPACING="0"
><TR
><TH
COLSPAN="3"
ALIGN="center"
>ATSプログラミングチュートリアル: </TH
></TR
><TR
><TD
WIDTH="10%"
ALIGN="left"
VALIGN="bottom"
><A
HREF="c93.html"
ACCESSKEY="P"
>Prev</A
></TD
><TD
WIDTH="80%"
ALIGN="center"
VALIGN="bottom"
></TD
><TD
WIDTH="10%"
ALIGN="right"
VALIGN="bottom"
><A
HREF="c194.html"
ACCESSKEY="N"
>Next</A
></TD
></TR
></TABLE
><HR
ALIGN="LEFT"
WIDTH="100%"></DIV
><DIV
CLASS="chapter"
><H1
><A
NAME="exceptions"
></A
>Chapter 5. Exceptions</H1
><P
>&#13;
While exceptions can be very useful in practice, it is also very common to
see code that misuses exceptions.

</P
><P
>&#13;Generally speaking, there are exceptions that are meant to be raised but
not captured for the purpose of aborting program execution, and there are
also exceptions (often declared locally) that are meant to be raised and
then captured so as to change the flow of program execution.  For instance,
the exception <FONT
COLOR="RED"
>ArraySubscriptExn</FONT
> is raised when out-of-bounds
array subscripting is detected at run-time.  Once it is raised,
<FONT
COLOR="RED"
>ArraySubscriptExn</FONT
> is usually not meant to be captured.  While
there is certainly nothing preventing a programer from writing code that
captures a raised <FONT
COLOR="RED"
>ArraySubscriptExn</FONT
>, a major concern is that
reasoning can become greatly complicated on code that does so.  In the
following presentation, I will soley focus on exceptions that are meant to
be raised and then captured.

</P
><P
>&#13;Let us now take a look at the following code that implements a function
for finding the rightmost element in a list that satisifes a given predicate:

<DIV
CLASS="informalexample"
><P
></P
><A
NAME="AEN167"
></A
><PRE
CLASS="programlisting"
>extern
fun{a:t@ype}
list_find_rightmost
  (List (a), (a) -&#60;cloref1&#62; bool): Option_vt (a)
//
implement{a}
list_find_rightmost
  (xs, pred) = let
//
fun aux
(
  xs: List(a)
) : Option_vt (a) =
  case+ xs of
  | nil () =&#62; None_vt ()
  | cons (x, xs) =&#62; let
      val res = aux (xs)
    in
      case+ res of
      | Some_vt _ =&#62; res
      | ~None_vt () =&#62;
          if pred (x) then Some_vt (x) else None_vt ()
        // end of [None]
    end (* end of [cons] *)
//
in
  aux (xs)
end // end of [list_find_rightmost]
</PRE
><P
></P
></DIV
>


Suppose that <FONT
COLOR="RED"
>list_find_rightmost</FONT
> is called on a list xs of
length N (for some large natural number N) and a predicate pred. The
evaluation of this call leads to a call to the inner function
<FONT
COLOR="RED"
>aux</FONT
>, which in turn generates N additional recursive calls to
<FONT
COLOR="RED"
>aux</FONT
>. Assume that only the last element of xs satisfies the
predicate pred. Then there are still N-1 call frames for <FONT
COLOR="RED"
>aux</FONT
> on
the call stack when the rightmost element satisfying pred is found, and
these frames need to be unwinded <SPAN
CLASS="emphasis"
><I
CLASS="emphasis"
>one-by-one</I
></SPAN
> before the found
element can be returned to the original call to
<FONT
COLOR="RED"
>list_find_rightmost</FONT
>. This form of inefficiency is eliminated in
the following exception-based implementation of
<FONT
COLOR="RED"
>list_find_rightmost</FONT
>:

<DIV
CLASS="informalexample"
><P
></P
><A
NAME="AEN176"
></A
><PRE
CLASS="programlisting"
>&#13;implement{a}
list_find_rightmost
  (xs, pred) = let
//
exception Found of (a)
//
fun aux
(
  xs: List(a)
) : void =
  case+ xs of
  | nil () =&#62; ()
  | cons (x, xs) =&#62; let
      val () = aux (xs)
    in
      if pred (x) then $raise Found(x) else ()
    end (* end of [cons] *)
//
in
//
try let
  val () = aux (xs)
in
  None_vt ()
end with
  | ~Found(x) =&#62; Some_vt (x)
//
end // end of [list_find_rightmost]
</PRE
><P
></P
></DIV
>


When a try-with-expression is evaluated, a label is created for the portion
of the call stack needed to evaluate the clauses (often referred to as
exception-handlers) following the keyword <FONT
COLOR="RED"
>with</FONT
>, and this label
is then pushed onto a designated global stack.  When an exception is
raised, the labels on the global stack are tried one-by-one until the
raised exception is captured by an exception-handler (that is, the value
representing the exception matches the pattern guard of the
exception-handler) or the current program evaluation aborts. The above
exception-based implementation of <FONT
COLOR="RED"
>list_find_rightmost</FONT
> uses a
raised exception to carry the element found during a recursive call to
<FONT
COLOR="RED"
>aux</FONT
> so that this element can be returned in a single jump to
the original call to <FONT
COLOR="RED"
>list_find_rightmost</FONT
>, bypassing all the
intermediate call frames (for recursive calls to <FONT
COLOR="RED"
>aux</FONT
>) on the
call stack. In general, the range between the point where an exception
is raised and the point where the raised exception is captured should span
multiple call frames. If not, then the use of exception may be questionable.

</P
><P
>&#13;The implementation of the run-time support for exceptions in ATS makes use
of the function <FONT
COLOR="RED"
>alloca</FONT
> declared in <TT
CLASS="filename"
>alloca.h</TT
> and
the functions <FONT
COLOR="RED"
>setjmp</FONT
> and <FONT
COLOR="RED"
>longjmp</FONT
> declared in
<TT
CLASS="filename"
>setjmp.h</TT
>. If <B
CLASS="command"
>gcc</B
> or <B
CLASS="command"
>clang</B
> is used to
compile the C code generated from ATS source, one can pass the flag
-D_GNU_SOURCE so as to make sure that the header file <TT
CLASS="filename"
>alloca.h</TT
>
is properly included.

</P
><P
>&#13;Please find <A
HREF="http://www.ats-lang.org/DOCUMENT/ATS2TUTORIAL/CODE/chap_exception.dats"
TARGET="_top"
>on-line</A
> the entirety of the
code used in this chapter.
</P
></DIV
><DIV
CLASS="NAVFOOTER"
><HR
ALIGN="LEFT"
WIDTH="100%"><TABLE
SUMMARY="Footer navigation table"
WIDTH="100%"
BORDER="0"
CELLPADDING="0"
CELLSPACING="0"
><TR
><TD
WIDTH="33%"
ALIGN="left"
VALIGN="top"
><A
HREF="c93.html"
ACCESSKEY="P"
>Prev</A
></TD
><TD
WIDTH="34%"
ALIGN="center"
VALIGN="top"
><A
HREF="index.html"
ACCESSKEY="H"
>Home</A
></TD
><TD
WIDTH="33%"
ALIGN="right"
VALIGN="top"
><A
HREF="c194.html"
ACCESSKEY="N"
>Next</A
></TD
></TR
><TR
><TD
WIDTH="33%"
ALIGN="left"
VALIGN="top"
>Fixity Declarations</TD
><TD
WIDTH="34%"
ALIGN="center"
VALIGN="top"
><A
HREF="p37.html"
ACCESSKEY="U"
>Up</A
></TD
><TD
WIDTH="33%"
ALIGN="right"
VALIGN="top"
>References</TD
></TR
></TABLE
></DIV
></BODY
></HTML
>